FHEM/60_uba.pm

```diff
@@ -550,187 +240,266 @@ sub uba_Get($@) {
-    readingsSingleUpdate( $hash, "lastUpdate".$param->{type}, FmtDateTime($lastdata), 0 ) if(AttrVal($name, "showTimeReadings", 0) eq 1);
-  } else {
-    my $twentyfour = int(time)-(24*60*60);
-    readingsSingleUpdate( $hash, ".lastUpdate".$param->{type}, $twentyfour, 0 ) if($lastdata < $twentyfour);
-    Log3 $name, 2, "Skipping missing readings before ".FmtDateTime($twentyfour) if($lastdata < $twentyfour);
-  }
+    # Save Luftqualitätsindex_name
+    readingsBulkUpdateIfChanged($hash,"luftqualitaetsindex_name",$airquality{$json->{"data"}->{$station}->{$datetimekey}[1]});
+    $hash->{CHANGETIME}[$changeindex++] = FmtDateTime($measured_datetime_timestamp);
+  
+    # Search and work on each specific available dataset
+    my $dataset_size = @{$json->{"data"}->{$station}->{$datetimekey}};
+    Log3 $name, 5, "x dataset_size: $dataset_size";
+    for ( my $i = 3; $i < $dataset_size; $i++ ) {
+      Log3 $name, 4, "-$i 0 component ID: ".$i;
+      Log3 $name, 4, "--$i 0 component: ".$json->{"data"}->{$station}->{$datetimekey}[$i][0];
+      Log3 $name, 4, "--$i 0 component_name: ".$components{$json->{"data"}->{$station}->{$datetimekey}[$i][0]};
+      Log3 $name, 4, "--$i 1 data: ".$json->{"data"}->{$station}->{$datetimekey}[$i][1];
+
+      # Save Datasets
+      readingsBulkUpdateIfChanged($hash, $components{$json->{"data"}->{$station}->{$datetimekey}[$i][0]}, $json->{"data"}->{$station}->{$datetimekey}[$i][1]);
+      $hash->{CHANGETIME}[$changeindex++] = FmtDateTime($measured_datetime_timestamp);
+    }
+
+    # Updaten der .lastUpdate
+    if( $measured_datetime_timestamp > $lastupdate_timestamp ){
+      # Er soll nur den höchsten Wert schreiben
+      Log3 $name, 4, "Writing new lastUpdate ($measured_datetime_timestamp)";
+      readingsBulkUpdateIfChanged($hash,".lastUpdate",$measured_datetime_timestamp, 0);
+      readingsBulkUpdateIfChanged($hash,"lastUpdate",FmtDateTime($measured_datetime_timestamp), 0) if(AttrVal($name, "showTimeReadings", 0) eq 1);
+    }
 
+    # End data update
+    readingsEndUpdate($hash, 1);
+  }
+  
+  # Set timer
   my $nextupdate = gettimeofday()+$hash->{helper}{INTERVAL};
   RemoveInternalTimer($hash, "uba_GetUpdate");
-  InternalTimer($nextupdate, "uba_GetUpdate", $hash, 1);
+  InternalTimer($nextupdate, "uba_GetUpdate", $hash);
+
+  # set program state/status
+  readingsSingleUpdate($hash,'state','done',0);
+  readingsSingleUpdate($hash,'Quellenangabe','Umweltbundesamt mit Daten der Messnetze der Länder und des Bundes',0);
+  Log3 $name, 3, "UBA: Done loading all data";
 
   return undef;
 }
 
-sub uba_Attr(@)
+
+sub Attr(@)
 {
   my ($cmd, $device, $attribName, $attribVal) = @_;
   my $hash = $defs{$device};
 
   $attribVal = "" if (!defined($attribVal));
-
-  if($cmd eq "set" && $attribName eq "userPassODL")
-  {
-    $attr{$device}{"userPassODL"} = uba_encrypt($attribVal);
-  }
   
   return undef;
 }
 
-sub uba_DbLog_splitFn($) {
+
+sub DbLog_splitFn($) {
```

Comment:
Der ganze `split`-Block is komisch, die Zuweisung könntest du auch direkt beim `split` machen.